File: ios/Orcha/Screens/TaskThreadScreen.swift
import SwiftUI
import UIKit

// Responsibility: Full task-thread transcript, paging, and human message composer.

struct TaskThreadScreen: View {
    @Environment(AppModel.self) private var model
    @Environment(\.palette) private var p
    let taskId: String

    @State private var draft = ""
    @State private var pendingSend: String?
    /// GH #140 — a tapped task-id link pushes here, in addition to the tab's own
    /// `WorkspaceRoute.task` destination; both target the same `TaskDetailScreen`.
    @State private var linkedTaskId: String?

    private var task: TaskDto? { model.snapshot?.tasks.first { $0.id == taskId } }
    private var assignee: String? { task?.assignees.first ?? task?.ownerAlias }

    var body: some View {
        ScrollViewReader { proxy in
            ScrollView {
                VStack(alignment: .leading, spacing: 8) {
                    // Issue 4: "Load earlier" reveals the previous keyset page at the TOP; it
                    // prepends older messages and must NOT scroll the view to the bottom.
                    if model.threadHasMore {
                        Button {
                            Task { await model.loadEarlierThreadMessages(taskId) }
                        } label: {
                            if model.threadLoadingEarlier {
                                ProgressView().frame(maxWidth: .infinity)
                            } else {
                                Text("Load earlier messages")
                                    .font(.system(size: 12, weight: .bold))
                                    .foregroundStyle(p.accent)
                                    .frame(maxWidth: .infinity)
                            }
                        }
                        .buttonStyle(.plain)
                        .padding(.vertical, 4)
                        .disabled(model.threadLoadingEarlier)
                    }
                    if model.taskMessages.isEmpty, pendingSend == nil {
                        OrchaCard {
                            Text("No messages yet — say hi to \(assignee ?? "the assignee").")
                                .foregroundStyle(p.muted)
                        }
                    }
                    ForEach(Array(model.taskMessages.enumerated()), id: \.offset) { _, msg in
                        threadBubble(msg)
                    }
                    if let unsent = pendingSend {
                        VStack(alignment: .trailing, spacing: 2) {
                            Bubble(.mine, unsent)
                            if !model.actionInFlight {
                                Button("Not sent · Tap to retry") { send(unsent) }
                                    .buttonStyle(.plain)
                                    .font(.system(size: 11, weight: .bold))
                                    .foregroundStyle(p.danger)
                            }
                        }
                        .frame(maxWidth: .infinity, alignment: .trailing)
                    } else if let error = model.error {
                        Banner(kind: .danger, text: error)
                    }
                    Color.clear.frame(height: 1).id("bottom")
                }
                .padding(16)
            }
            // Scroll to bottom only when the NEWEST message changes (a new/sent message) or a
            // pending bubble appears — never on a "Load earlier" prepend (which changes the top).
            .onChange(of: model.taskMessages.last?.messageId) {
                withAnimation { proxy.scrollTo("bottom", anchor: .bottom) }
            }
            .onChange(of: pendingSend) {
                withAnimation { proxy.scrollTo("bottom", anchor: .bottom) }
            }
            .onReceive(NotificationCenter.default.publisher(for: UIResponder.keyboardWillShowNotification)) { _ in
                withAnimation { proxy.scrollTo("bottom", anchor: .bottom) }
            }
            .onAppear { proxy.scrollTo("bottom", anchor: .bottom) }
        }
        .safeAreaInset(edge: .bottom) { composer }
        .navigationTitle("Thread")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .principal) {
                VStack(spacing: 1) {
                    Text("Thread")
                        .font(.system(size: 15, weight: .semibold))
                        .foregroundStyle(p.text)
                    if let title = task?.title {
                        Text(title)
                            .font(.system(size: 11))
                            .foregroundStyle(p.muted)
                            .lineLimit(1)
                    }
                }
            }
        }
        .task { await model.loadTaskDetail(taskId) }
        .refreshable { await model.loadTaskDetail(taskId) }
        .navigationDestination(item: $linkedTaskId) { TaskDetailScreen(taskId: $0) }
    }

    @ViewBuilder
    private func threadBubble(_ msg: TaskMessageDto) -> some View {
        let tasks = model.snapshot?.tasks ?? []
        if msg.authorId == nil, !msg.isHuman {
            Bubble(.system, msg.body, tasks: tasks, onTapTask: { linkedTaskId = $0 })
        } else if msg.authorId != nil, msg.authorId == model.humanId {
            Bubble(.mine, msg.body, time: MobileUx.agoLabel(msg.createdAt), tasks: tasks, onTapTask: { linkedTaskId = $0 })
        } else {
            Bubble(
                .theirs, msg.body,
                author: msg.authorAlias ?? (msg.isHuman ? "human" : "agent"),
                time: MobileUx.agoLabel(msg.createdAt),
                tasks: tasks, onTapTask: { linkedTaskId = $0 }
            )
        }
    }

    /// `.composer` — rounded field + circular send button.
    private var composer: some View {
        HStack(alignment: .bottom, spacing: 8) {
            TextField("Message \(assignee ?? "the thread")…", text: $draft, axis: .vertical)
                .lineLimit(1...4)
                .font(.system(size: 14.5))
                .padding(.horizontal, 14)
                .padding(.vertical, 10)
                .background(p.surface2, in: RoundedRectangle(cornerRadius: 20))
                .overlay(RoundedRectangle(cornerRadius: 20).strokeBorder(p.border2, lineWidth: 1))
            Button(action: sendDraft) {
                Image(systemName: "arrow.up.circle.fill")
                    .font(.system(size: 32))
                    .foregroundStyle(canSend ? p.accent : p.faint)
            }
            .buttonStyle(.plain)
            .disabled(!canSend)
            .accessibilityLabel("Send")
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(p.bg)
    }

    private var canSend: Bool {
        !draft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && !model.actionInFlight
    }

    private func sendDraft() {
        let text = draft.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        draft = ""
        send(text)
    }

    /// A send that errors keeps its text as an unsent bubble with a retry chip.
    private func send(_ text: String) {
        pendingSend = text
        Task {
            if await model.sendTaskMessage(taskId, body: text) {
                pendingSend = nil
            }
        }
    }
}

/* ---------- flow 06 R2 — run detail: mono log, pin-to-bottom, stop-run ---------- */

/// Flow 06 R2 — run detail: header + stop-run, terminal banner, and the streaming
/// mono log filling the remaining space with pragmatic pin-to-bottom tracking.

File: android/app/src/main/java/io/openorcha/mobile/ui/screens/TaskThreadScreen.kt
package io.openorcha.mobile.ui.screens

/** Owns the paged task conversation and its message composer. */

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.consumeWindowInsets
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.isImeVisible
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import io.openorcha.mobile.ui.icons.OrchaIcons
import kotlinx.coroutines.launch
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import io.openorcha.mobile.data.RunDto
import io.openorcha.mobile.data.TaskDto
import io.openorcha.mobile.data.TaskMessageDto
import io.openorcha.mobile.domain.MobileUx
import io.openorcha.mobile.ui.OrchaUiState
import io.openorcha.mobile.ui.components.Avatar
import io.openorcha.mobile.ui.components.AvatarSize
import io.openorcha.mobile.ui.components.Banner
import io.openorcha.mobile.ui.components.BannerKind
import io.openorcha.mobile.ui.components.Bubble
import io.openorcha.mobile.ui.components.BubbleKind
import io.openorcha.mobile.ui.components.DangerTonalButton
import io.openorcha.mobile.ui.components.FeedRow
import io.openorcha.mobile.ui.components.MetaTag
import io.openorcha.mobile.ui.components.OrchaCard
import io.openorcha.mobile.ui.components.OrchaField
import io.openorcha.mobile.ui.components.PrimaryButton
import io.openorcha.mobile.ui.components.SectionH
import io.openorcha.mobile.ui.components.StatusDomain
import io.openorcha.mobile.ui.components.StatusPill
import io.openorcha.mobile.ui.components.TonalButton
import io.openorcha.mobile.ui.theme.MonoSmStyle
import io.openorcha.mobile.ui.theme.MonoStyle
import io.openorcha.mobile.ui.theme.Orcha

/* =============================================================================
   Flow 05 — Task detail + thread. Flow 06 — worker runs + streaming log.
   ============================================================================= */

@OptIn(ExperimentalMaterial3Api::class, ExperimentalLayoutApi::class)
@Composable
fun TaskThreadScreen(
    state: OrchaUiState,
    onBack: () -> Unit,
    onRefresh: () -> Unit,
    onSendMessage: (String) -> Unit,
    onLoadEarlier: () -> Unit,
    onOpenTask: (String) -> Unit,
) {
    val p = Orcha.palette
    val task = state.selectedTask
    var draft by remember { mutableStateOf("") }
    var pendingSend by remember { mutableStateOf<String?>(null) }
    // a send that errored keeps its text as an unsent bubble with a retry chip
    val unsent = if (state.error != null) pendingSend else null
    val listState = rememberLazyListState()
    val imeVisible = WindowInsets.isImeVisible
    // issue 2: keep the newest messages in view when the keyboard opens or a message lands.
    // Keyed on the NEWEST message's identity (same expression as the item keys), not the list
    // size — a "Load earlier" prepend grows the size but leaves the newest message unchanged,
    // so the effect stays put and LazyColumn's key-based anchoring holds the viewport at the seam.
    val newestMessageKey = state.taskMessages.lastOrNull()?.let { it.messageId ?: "${it.createdAt}-${it.body.hashCode()}" }
    LaunchedEffect(newestMessageKey, imeVisible) {
        val last = listState.layoutInfo.totalItemsCount - 1
        if (last >= 0 && (imeVisible || state.taskMessages.isNotEmpty())) listState.animateScrollToItem(last)
    }
    Scaffold(
        containerColor = Color.Transparent,
        topBar = {
            TopAppBar(
                colors = TopAppBarDefaults.topAppBarColors(containerColor = Color.Transparent),
                title = {
                    Column {
                        Text("Thread", style = MaterialTheme.typography.titleMedium)
                        Text(task?.title ?: "", style = MaterialTheme.typography.bodyMedium, color = p.muted, maxLines = 1, overflow = TextOverflow.Ellipsis)
                    }
                },
                navigationIcon = { IconButton(onClick = onBack) { Icon(OrchaIcons.ArrowBack, "Back") } },
                actions = { IconButton(onClick = onRefresh) { Icon(OrchaIcons.Refresh, "Refresh") } },
            )
        },
    ) { padding ->
        // issue 2: consumeWindowInsets stops imePadding re-adding the nav-bar inset the
        // Scaffold padding already applied (with adjustResize, that was the visible gap)
        Column(Modifier.fillMaxSize().padding(padding).consumeWindowInsets(padding).imePadding()) {
            LazyColumn(
                modifier = Modifier.weight(1f),
                state = listState,
                contentPadding = PaddingValues(16.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                // issue 4: keyset "Load earlier" — older pages prepend above (web reveal affordance)
                if (state.threadHasMore) {
                    item(key = "load-earlier") {
                        androidx.compose.material3.TextButton(
                            onClick = onLoadEarlier,
                            enabled = !state.threadLoadingEarlier,
                            modifier = Modifier.fillMaxWidth(),
                        ) {
                            Text(
                                if (state.threadLoadingEarlier) "Loading…" else "Load earlier messages",
                                color = p.accent, fontWeight = FontWeight.W700,
                            )
                        }
                    }
                }
                if (state.taskMessages.isEmpty()) {
                    item {
                        OrchaCard {
                            Text(
                                "No messages yet — say hi to ${task?.assignees?.firstOrNull() ?: "the assignee"}.",
                                color = p.muted,
                            )
                        }
                    }
                }
                items(state.taskMessages, key = { it.messageId ?: "${it.createdAt}-${it.body.hashCode()}" }) { msg ->
                    ThreadBubble(msg, state.selectedContainer?.humanAgentId, state.snapshot?.tasks.orEmpty(), onOpenTask)
                }
                unsent?.let { text ->
                    item {
                        Column(horizontalAlignment = Alignment.End, modifier = Modifier.fillMaxWidth()) {
                            Bubble(BubbleKind.Mine, text)
                            Text(
                                "Not sent · Tap to retry",
                                style = MaterialTheme.typography.labelMedium,
                                color = p.danger,
                                modifier = Modifier
                                    .padding(top = 2.dp)
                                    .clickable { onSendMessage(text) },
                            )
                        }
                    }
                }
                if (unsent == null) state.error?.let { item { Banner(BannerKind.Danger, it) } }
            }
            // `.composer` — rounded field + round send button
            Row(
                Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 8.dp),
                verticalAlignment = Alignment.Bottom,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                OrchaField(
                    draft, { draft = it },
                    modifier = Modifier.weight(1f),
                    placeholder = "Message ${task?.assignees?.firstOrNull() ?: "the thread"}…",
                    maxLines = 4,
                )
                IconButton(
                    onClick = { pendingSend = draft.trim(); onSendMessage(draft.trim()); draft = "" },
                    enabled = draft.isNotBlank() && !state.actionInFlight,
                    colors = androidx.compose.material3.IconButtonDefaults.iconButtonColors(
                        containerColor = p.accent, contentColor = p.accentInk,
                        disabledContainerColor = p.accent.copy(alpha = 0.4f), disabledContentColor = p.accentInk,
                    ),
                ) { Icon(OrchaIcons.Send, "Send") }
            }
        }
    }
}

@Composable
private fun ThreadBubble(msg: TaskMessageDto, humanId: String?, tasks: List<TaskDto>, onOpenTask: (String) -> Unit) {
    val mine = msg.authorId != null && msg.authorId == humanId
    val system = msg.authorId == null && !msg.isHuman
    when {
        system -> Bubble(BubbleKind.System, msg.body, tasks = tasks, onOpenTask = onOpenTask)
        mine -> Bubble(BubbleKind.Mine, msg.body, time = MobileUx.agoLabel(msg.createdAt), tasks = tasks, onOpenTask = onOpenTask)
        else -> Bubble(
            BubbleKind.Theirs, msg.body,
            author = msg.authorAlias ?: if (msg.isHuman) "human" else "agent",
            time = MobileUx.agoLabel(msg.createdAt),
            tasks = tasks, onOpenTask = onOpenTask,
        )
    }
}

/* ---------- flow 06 R2 — run detail: mono log, pin-to-bottom, stop-run ---------- */
